### PR TITLE
take out macros from loops

### PR DESCRIPTION
**Pull request by jangorecki**
avoid overhead of R macros
no logic change, only take out macros out of the loops, and few DATAPTR removal

### CODE REVIEW

**Comment by jangorecki** [comment]
I removed some `DATAPTR` in `bmerge` also. They were used in `twiddle`, as a result they couldn't be `const double *` but just `double *`, address https://github.com/Rdatatable/data.table/issues/3301
some macros in bmerge were not removed as I want this PR to be very simple, and logic there would need to be rewritten more heavily.

**Comment by jangorecki** [comment]
not sure how much noise is in this measurement, but `test.data.table` timings on
master: 58.5s
c-loops: 49.5s

File: src/dogroups.c [hunk 1]
```diff
@@ -177,45 +180,45 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
...
         if (LOGICAL(verbose)[0]) { tblock[0] += clock()-tstart; nblock[0]++; }
       } else {
         // Fairly happy with this block. No need for SET_* here. See comment above.
-        for (k=0; k<grpn; k++) INTEGER(I)[k] = INTEGER(order)[ INTEGER(starts)[i]-1 + k ];
+        for (k=0; k<grpn; k++) iI[k] = iorder[ istarts[i]-1 + k ];
         for (j=0; j<length(SDall); j++) {
           size_t size = SIZEOF(VECTOR_ELT(SDall,j));
           target = VECTOR_ELT(SDall,j);
           source = VECTOR_ELT(dt,INTEGER(dtcols)[j]-1);
-          int *Id = INTEGER(I);
```
Comment (jangorecki) [inline src/dogroups.c]: this was initialized sooner, and already used sooner

File: src/dogroups.c [hunk 2]
```diff
@@ -177,45 +180,45 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
...
-            memcpy((char *)DATAPTR(VECTOR_ELT(SDall,j)),  // direct memcpy best here, for usually large size groups. by= each row is slow and not recommended anyway, so we don't mind there's no switch here for grpn==1
+            memcpy((char *)DATAPTR(VECTOR_ELT(SDall,j)),  // direct memcpy best here, for usually large size groups. by= each row is slow and not recommended anyway, so we don't mind there's no switch here for grpn==1 // TODO remove DATAPTR
                    (char *)DATAPTR(VECTOR_ELT(dt,INTEGER(dtcols)[j]-1))+rownum*size,
                    grpn*size);
             // SD is our own alloc'd memory, and the source (DT) is protected throughout, so no need for SET_* overhead
           }
           for (j=0; j<length(xSD); j++) {
             size_t size = SIZEOF(VECTOR_ELT(xSD,j));
-            memcpy((char *)DATAPTR(VECTOR_ELT(xSD,j)),  // ok use of memcpy. Loop'd through columns not rows
+            memcpy((char *)DATAPTR(VECTOR_ELT(xSD,j)),  // ok use of memcpy. Loop'd through columns not rows // TODO remove DATAPTR
```
Comment (jangorecki) [inline src/dogroups.c]: in cases like this we have to `switch` and case by case? or is there any better way?

Comment (mattdowle) [reply src/dogroups.c]: maybe call `memrecycle`

**Comment by mattdowle** [comment]
I added a check to setcolorder at C level that it has a strict permutation. In future the (expensive) `anyDuplicated` at R level setcolcorder could be removed and left to C level, with the C level errors embellished and tested with coverage of course.
As I was going through checking I started to move variables down to local scope while I was there (particularly loop counters) and changed i++ to ++i.  We might as well give all compilers on all platforms, even old ones, the best chance:  https://www.quora.com/Do-you-use-i++-or-++i-in-a-for-loop-And-why/answer/Matt-Parsons-12
I removed curly braces around case statements where no local variables are created for that case.  When I see a curly after a case, I think "ok, local variable is here, I'm expecting to see one".  That's the only reason a case needs curlies otherwise the local variable declaration doesn't compile.

File: src/fmelt.c [hunk 3]
```diff
@@ -478,29 +478,19 @@ SEXP getvaluecols(SEXP DT, SEXP dtnames, Rboolean valfactor, Rboolean verbose, s
...
+          for (int k=0; k<thislen; ++k)
             itarget[counter + k] = ithiscol[ithisidx[k]-1];
         } else {
           memcpy(itarget + j*data->nrow, ithiscol, data->nrow*size);
         }
-      }
-        break;
-      default : error("Unknown column type '%s' for column '%s'.", type2char(TYPEOF(thiscol)), CHAR(STRING_ELT(dtnames, INTEGER(thisvaluecols)[i]-1)));
+      } break;
+      default :
```
Comment (jangorecki) [inline src/fmelt.c]: I would keep it in same line, so if you collapse `{` of switch then every case, including default, takes only single line

File: src/bmerge.c [hunk 4]
```diff
@@ -298,7 +298,7 @@ void bmerge_r(int xlowIn, int xuppIn, int ilowIn, int iuppIn, int col, int thisg
     // ilow and iupp now surround the group in ic, too
   }
     break;
-  case STRSXP : {
```
Comment (jangorecki) [inline src/bmerge.c]: this `{` was here for consistency to other cases, and nice collapsing code

File: src/dogroups.c [hunk 5]
```diff
@@ -48,8 +48,8 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
   iSD = PROTECT(findVar(install(".iSD"), env)); protecti++; // 1-row and possibly no cols (if no i variables are used via JIS)
   xSD = PROTECT(findVar(install(".xSD"), env)); protecti++;
   R_len_t maxGrpSize = 0;
-  const int *ilens = INTEGER(lens);
-  for (R_len_t i=0; i<LENGTH(lens); i++) {
+  const int *ilens = INTEGER(lens), n=LENGTH(lens);
+  for (R_len_t i=0; i<n; ++i) {
```
Comment (jangorecki) [inline src/dogroups.c]: @mattdowle in this commit many changes from `i++` to `++i` or reverse, but at the same time no logic change, what is the reason for that?

File: src/bmerge.c (hunk 4 above)
Comment (mattdowle) [reply src/bmerge.c]: by collapsing do mean in editor?  If we add `{` to all `case:` I don't think that's very standard.  I wrote my thinking here:  https://github.com/Rdatatable/data.table/pull/3610#issuecomment-497099447

File: src/dogroups.c (hunk 5 above)
Comment (mattdowle) [reply src/dogroups.c]: Did you see https://github.com/Rdatatable/data.table/pull/3610#issuecomment-497099447?

Comment (jangorecki) [reply src/dogroups.c]: found answer in PR comment

File: src/bmerge.c (hunk 4 above)
Comment (jangorecki) [reply src/bmerge.c]: yes, collapsing code in IDE, makes it very convenient

File: src/fmelt.c (hunk 3 above)
Comment (mattdowle) [reply src/fmelt.c]: I like long lines as you know, but this seems a step too far to put the error on the same line as the default:  For the reason that the other cases in this switch aren't like that due to being bigger.  When all the cases are one short line, then yes but not in this case. Consistency within the switch.

Comment (jangorecki) [reply src/fmelt.c]: then what can help here is
```
default : {
  error("a");
}
```
collapse-able to single line

**Pull request merged**